### PR TITLE
added deserializing functionality

### PR DESCRIPTION
**Pull request by ManfredTremmel**
I've tried to use problem-spring-web to delegate excpetions between spring-boot based microservices, but wasn't able to deserialize the serialized exceptions. With the made changes it works.

### CODE REVIEW

**Comment by whiskeysierra** [comment]
@ManfredTremmel I don't quite understand the problem that you're having. The scope of this project is to produce problems as part of a Spring Web MVC application. Deserialization is provided by zalando/problem already. Have you read the [Handling problems](https://github.com/zalando/problem#handling-problems) section in the README?

**Comment by whiskeysierra** [comment]
@ManfredTremmel I suggest you check the following list of suggestions that may be giving you some issues when deserializing proiblems:
- make sure you understand the difference between zalando/problem (de/serialization of problems) and zalando/problem-spring-web (converting exceptions in Spring Web MVC to problems)
- make sure you register the `ProblemModule` in your ObjectMapper
- make sure you register custom sub types, e.g. `ConstraintViolationProblem`.
- you don't need `@JsonProperty("..")` on constructor parameters if you enable `-parameters` in [javac](http://docs.oracle.com/javase/8/docs/technotes/tools/windows/javac.html)

**Comment by ManfredTremmel** [comment]
@whiskeysierra  thank you for response.  I've tried hard to follow your description, but I wasn't able to get it run. Ok, registering ProblemModule and sub type ConstraintViolationProblem enables me to remove AbstractPolymorphicThrowableProblem, but Deserializing "status" always fails, because it's not numeric. I have to take a look, why "status" contains the message and not the http state in our service.
I'll let you know tomorow, if I can find a solution for this problem, then I can close this request.

**Comment by whiskeysierra** [comment]
Are you using the ProblemModule also on your other service? According to the RFC the status field should be an integer.

**Comment by ManfredTremmel** [comment]
At the moment the first usage is a persistens microservice based on spring data jpa which uses the exception handler. The [test data](https://github.com/ManfredTremmel/problem-spring-web/blob/master/src/test/resources/org/zalando/problem/deserialize/ValidationError.json) I've added has been created by this service.  I havn't found out today, why the status is serialized as enum name and not as http state number, but I'm on it.
When everything works, I'll try to use it also on GWT-Frontends, at the moment I'm using a limited [Exception Handler](https://github.com/ManfredTremmel/gwt-bean-validators-spring-gwtp/blob/master/src/main/java/de/knightsoftnet/validators/server/controller/RestErrorHandler.java) and a [Rest Callback](https://github.com/ManfredTremmel/gwt-bean-validators-spring-gwtp/blob/master/src/main/java/de/knightsoftnet/validators/client/rest/helper/AbstractRestCallback.java)  to deserialize validation results.

**Comment by whiskeysierra** [comment]
> I havn't found out today, why the status is serialized as enum name and not as http state number, but I'm on it.

Your object mapper is not using the ProblemModule. You can see that by the presence of properties like `localizedMessage`, `stackTrace`, `message`, `surpressed`...

I'd suggest to verify the Jackson setup in that service.

**Comment by ManfredTremmel** [comment]
I've found the problem in our service. Spring data uses the webmvc converters, so I had to modify the mappers there:

```
@Configuration
public class WebMvcConfiguration extends WebMvcConfigurationSupport {
	@Override
	protected void extendMessageConverters(List<HttpMessageConverter<?>> converters) {
		super.extendMessageConverters(converters);
		for (final HttpMessageConverter<?> converter : converters) {
			if (converter instanceof MappingJackson2HttpMessageConverter) {
				final ObjectMapper objectMapper = ((MappingJackson2HttpMessageConverter) converter).getObjectMapper();
		        objectMapper.registerModule(new Jdk8Module());
		        objectMapper.registerModule(new ProblemModule());
		        objectMapper.registerSubtypes(ConstraintViolationProblem.class);
			}
		}
	}
}

```

What I'm still unhappy about is the missing @JsonProperty("..") annotations, I can't set the -parameter javac option inside eclipse, so I can't run the tests in the IDE. Would it be possible to add this annotations? I can send another change request if you want.

**Comment by whiskeysierra** [comment]
> What I'm still unhappy about is the missing @JsonProperty("..") annotations, I can't set the -parameter javac option inside eclipse, so I can't run the tests in the IDE. Would it be possible to add this annotations?

That actually makes sense. Not having annotations allows clients to customize e.g. the naming strategy, but this is actually not desired here, since the property names are strictly defined by the RFC. I'll add annotations.

**Comment by whiskeysierra** [comment]
@ManfredTremmel After checking it again, there should be no need to use `-parameters` or `@JsonProperty` at all. We either already have the annotations in place (check [Problem](https://github.com/zalando/problem/blob/master/jackson-datatype-problem/src/main/java/org/zalando/problem/ProblemMixIn.java#L21)) or the names are actually words that case-agnostic, e.g. *violations*, *message*, *field*, ...

Can you maybe describe what your actual problem is?

**Comment by ManfredTremmel** [comment]
The problem is, when I try to deserialize a ConstraintViolationProblem, a JsonMappingException is thrown, triggered by a IllegalArgumentException:

```
Caused by: java.lang.IllegalArgumentException: Argument #0 of constructor [constructor for org.zalando.problem.spring.web.advice.validation.ConstraintViolationProblem, annotations: {interface com.fasterxml.jackson.annotation.JsonCreator=@com.fasterxml.jackson.annotation.JsonCreator(mode=DEFAULT)}] has no property name annotation; must have name when multiple-parameter constructor annotated as Creator
	at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory._addDeserializerConstructors(BasicDeserializerFactory.java:511)
...
```

**Comment by whiskeysierra** [comment]
> when I try to deserialize a ConstraintViolationProblem

That is your problem right there. When you read the docs for that particular problem you find

> It’s not meant to be used for end-user input validation, but for client developer convenience.

Basically there is no point in deserializing it. It shouldn't be handled in the code directly. I'd suggest to **not** register ConstraintViolationProblem as a subtype in Jackson and just deserialize it into a `DefaultProblem`. Throwing and logging that will give you everything you need for analyzing/debugging.

**Comment by whiskeysierra** [comment]
I'll close the pull request. Comments on this are still welcome.
